src/uct/ib/rc/accel/rc_mlx5_ep.c

```diff
@@ -253,6 +254,122 @@ ucs_status_t uct_rc_mlx5_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
+
+    desc = ucs_mpool_get_inline(&iface->mlx5_common.dm.dm->mp);
+    if (ucs_unlikely((desc) == NULL)) {
+        /* in case if no resources available - fallback to bcopy */
+        UCT_RC_IFACE_GET_TX_DESC(rc_iface, &rc_iface->tx.mp, desc);
+        desc->super.handler = (uct_rc_send_handler_t)ucs_mpool_put;
+
+        memcpy(desc + 1, cache->out, hdr_len);
+        memcpy(UCS_PTR_BYTE_OFFSET(desc + 1, hdr_len), payload, length);
+
+        uct_rc_mlx5_txqp_bcopy_post(rc_iface, &ep->super.txqp, &ep->tx.wq,
+                                    MLX5_OPCODE_SEND, hdr_len + length,
+                                    0, 0, MLX5_WQE_CTRL_SOLICITED, 0, desc, desc + 1);
+        if (stat) {
+            UCT_TL_EP_STAT_OP(&ep->super.super, AM, BCOPY, length);
+        }
+    } else {
+        ucs_assert(desc->super.buffer != NULL);
+
+        uct_rc_mlx5_ep_copy_to_dm(cache, hdr_len, payload,
+                                  length, desc->super.buffer);
+
+        uct_rc_mlx5_txqp_bcopy_post(rc_iface, &ep->super.txqp, &ep->tx.wq,
+                                    MLX5_OPCODE_SEND, hdr_len + length,
+                                    0, 0, MLX5_WQE_CTRL_SOLICITED | MLX5_WQE_CTRL_CQ_UPDATE,
+                                    0, desc,
+                                    (void*)(desc->super.buffer - iface->mlx5_common.dm.dm->start_va));
+        if (stat) {
+            UCT_TL_EP_STAT_OP(&ep->super.super, AM, SHORT, hdr_len + length);
+        }
+    }
+    return UCS_OK;
+}
+
+static ucs_status_t UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_ep_am_short_dm(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
+                           const void *payload, unsigned length)
+{
+    uct_rc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_rc_mlx5_iface_t);
+    uct_rc_iface_t *rc_iface   = &iface->super;
+    uct_rc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_rc_mlx5_ep_t);
+    ucs_status_t status;
+    uct_rc_mlx5_dm_copy_data_t cache;
+
+    if (ucs_likely(sizeof(uct_rc_am_short_hdr_t) + length <= UCT_IB_MLX5_AM_MAX_SHORT(0))) {
+        return uct_rc_mlx5_ep_am_short_inline(tl_ep, id, hdr, payload, length);
+    }
+
+    UCT_CHECK_LENGTH(length + sizeof(uct_rc_am_short_hdr_t), 0,
+                     iface->mlx5_common.dm.seg_len, "am_short");
+    UCT_CHECK_AM_ID(id);
+    UCT_RC_CHECK_RES(rc_iface, &ep->super);
+    UCT_RC_CHECK_FC(rc_iface, &ep->super, id);
+
+    uct_rc_am_hdr_fill(&cache.am_hdr.rc_hdr, id);
+    cache.am_hdr.am_hdr = hdr;
+
+    status = uct_rc_mlx5_ep_short_dm(ep, &cache, sizeof(cache.am_hdr), payload, length, 1);
+
+    if (UCS_STATUS_IS_ERR(status)) {
+        return status;
+    }
+
+    UCT_RC_UPDATE_FC(rc_iface, &ep->super, id);
+    return UCS_OK;
+}
+#endif
+
+ucs_status_t
+uct_rc_mlx5_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
+                        const void *payload, unsigned length)
+{
+#if HAVE_IBV_EXP_DM
+    return uct_rc_mlx5_ep_am_short_dm(tl_ep, id, hdr, payload, length);
```

Comment:
can we unite the functions uct_rc_mlx5_ep_am_short() and uct_rc_mlx5_ep_am_short_dm() to reduce some code?

Reply:
done